CASMINST-4915 ncn-image-modification (#2000)
Cray-HPE/docs-csm: install/pre-installation.md

      install -y csm-testing goss-servers
   ```

1. (`pit#`) Get the artifact versions.

   ```bash
   kubernetes_version="$(find ${CSM_PATH}/images/kubernetes -name '*.squashfs' -exec basename {} .squashfs \; | awk -F '-' '{print $NF}')"
   ceph_version="$(find ${CSM_PATH}/images/storage-ceph -name '*.squashfs' -exec basename {} .squashfs \; | awk -F '-' '{print $NF}')"
   ```

> 

Author: Good question, very likely. I can have it exclude the `old/` dir.

> 

Author: Well, actually it won't.

The `ncn-image-modification` does not modify files in `${PITDATA}/images/`

1. (`pit#`) Copy the NCN images from the expanded tarball.

   > ***NOTE*** This hard-links the files to do this copy as fast as possible, as well as to mitigate space waste on the USB stick.

   ```bash
   rsync -rltDP --delete "${CSM_PATH}/images/kubernetes/" --link-dest="${CSM_PATH}/images/kubernetes/" "${PITDATA}/data/k8s/${kubernetes_version}"
   rsync -rltDP --delete "${CSM_PATH}/images/storage-ceph/" --link-dest="${CSM_PATH}/images/storage-ceph/" "${PITDATA}/data/ceph/${ceph_version}"
   ```

1. (`pit#`) Generate SSH Keys and invoke `ncn-image-modification.sh`:

   - Generate SSH Keys:

       > ***NOTE*** The code block below assumes an RSA key without a passphrase. If the administrator wants to customize this step they may do so

       ```bash
       ssh-keygen -N "" -t rsa
       ```

   - Run `ncn-image-modification.sh` from the CSM Tarball:

       ```bash

>        export SQUASHFS_ROOT_PW_HASH="$(awk -F':' /^root:/'{print $2}' < /etc/shadow)"

Author: @heemstra , I suppose I could either remove the empty line below.. or move this line into it's own step. Thoughts?

       "${PITDATA}/csm-${CSM_RELEASE}/ncn-image-modification.sh -p \
          -k /var/www/ephemeral/data/k8s/${kubernetes_version}/kubernetes-${kubernetes_version}.squashfs \
          -s /var/www/ephemeral/data/ceph/${kubernetes_version}/storage-ceph-${ceph_version}.squashfs"

>        ```

Reviewer: you're missing a `-d /root/.ssh` argument here.

      install -y csm-testing goss-servers
   ```

1. (`pit#`) Get the artifact versions.

   ```bash
   kubernetes_version="$(find ${CSM_PATH}/images/kubernetes -name '*.squashfs' -exec basename {} .squashfs \; | awk -F '-' '{print $NF}')"
   ceph_version="$(find ${CSM_PATH}/images/storage-ceph -name '*.squashfs' -exec basename {} .squashfs \; | awk -F '-' '{print $NF}')"
   ```

1. (`pit#`) Copy the NCN images from the expanded tarball.

   > ***NOTE*** This hard-links the files to do this copy as fast as possible, as well as to mitigate space waste on the USB stick.

   ```bash
   rsync -rltDP --delete "${CSM_PATH}/images/kubernetes/" --link-dest="${CSM_PATH}/images/kubernetes/" "${PITDATA}/data/k8s/${kubernetes_version}"
   rsync -rltDP --delete "${CSM_PATH}/images/storage-ceph/" --link-dest="${CSM_PATH}/images/storage-ceph/" "${PITDATA}/data/ceph/${ceph_version}"
   ```

1. (`pit#`) Generate SSH Keys and invoke `ncn-image-modification.sh`:

>    - Generate SSH Keys:

Reviewer 2: ```suggestion
   1. Generate SSH keys.
```

>        > ***NOTE*** The code block below assumes an RSA key without a passphrase. If the administrator wants to customize this step they may do so

Reviewer 2: ```suggestion
       > ***NOTE*** The code block below assumes there is an RSA key without a passphrase. This step can be customized to use a passphrase if desired.
```

       ```bash
       ssh-keygen -N "" -t rsa
       ```

>    - Export the password hash for `root`, this is needed by `ncn-image-modification.sh`:

Reviewer 2: ```suggestion
   1. Export the password hash for `root` that is needed for the `ncn-image-modification.sh` script:
```

       ```bash
       export SQUASHFS_ROOT_PW_HASH="$(awk -F':' /^root:/'{print $2}' < /etc/shadow)"
       ```

>    - Run `ncn-image-modification.sh` from the CSM Tarball:

Reviewer 2: ```suggestion
   1. Run `ncn-image-modification.sh` from the CSM tarball:
```
